### PR TITLE
More robust and flexible dataloader (bytes_dataloader deprecated)

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by iejMac** [comment]
## Notes
At a high level it looks like train_dataloader.py is the main file that initializes everything and creates the correct dataset and returns it. filter.py is nice - I like a dedicated script for filters we can apply to datasets that either filter based on the streams or metadata. We are still doing cut detection in the dataloader, not sure if we want this since we moved this to the dataset creation stage. Would reduce the amount of code here which I think is positive for a dataloader.

WebDatasetWithChagnedDecoder looks interesting although it seems to only be used when cut detection is happening which if we don't want will mean it won't be used anymore. Need to look into this more.

### train_dataloader.py:
- no PyTorch dataloader. Do we want that?
- cut detection is happening maybe we remove that since we added the subsampler at dataset creation phase
- need to figure out how to pass args in

## custom_wds.py:
- what are other uses of this? seems more general than just cut detection

### video_decode.py
- https://github.com/iejMac/video2dataset/blob/373a76938f7780bc41d483e18b183c430e0601c7/video2dataset/dataloader/video_decode.py#L85 - are the frames selected randomly?? how does this work?
- perhaps we can keep VideoDecoderCutDetection as an example of video decoders and how to write them in video2dataset. Maybe we also need to make a subsamplers type folder for this...

### filters.py
- maybe if we see more filters being added we can make this into a folder like with subsamplers so we can individually test the filters, not important now but something to think about for the future.

**Comment by iejMac** [comment]
# Notes v2

- Why do we need shardshuffle nad regular shuffle? Can't we just have one variable that decides if we shuffle at both levels? I don't see when you'd want one but not the other.
- Why the need for the VideoResizer? Why not just use torch vision transforms like in - https://github.com/iejMac/video2dataset/blob/5077529e3e82a01db76e7ac71d3c0a4bcf39c348/video2dataset/dataloader/transform.py#L155
- Do we need this? https://github.com/iejMac/video2dataset/blob/5077529e3e82a01db76e7ac71d3c0a4bcf39c348/video2dataset/dataloader/video_decode.py#L131 Can't we just read this off from the frames in the VideoResizer?

**Comment by iejMac** [comment]
ok lint is finally fixed

**Comment by iejMac** [comment]
One those questions are answered I'd also like to merge the bytes dataloader into a regular WebDataset with a bytes video decoder

**Comment by ablattmann** [comment]
Regarding your questions:
* yes we can skip shardshuffle, having one var for this is better 
* As far as I can see, your transform function is randomly cropping every individual frame independently of the others (is this the case?). This could lead to bad temporal consistency, since frames are not aligned then.  Apart from that, the Resizer offers a couple more combinations of resizing/cropping combinations than the `video_transform()` function, as far as I could see. But if you think that these are not needed, we can also remove it.
* Reading from `VideoResizer`? --> yes, we can do this.

**Comment by iejMac** [comment]
Ok all issues with current code have been resolved to my best knowledge. Going to try to deprecate the bytes dataloader now and then merge
